src/Aqovia.Utilities.SagaMachine/StatePersistance/RedisKeyValueStore.cs

```diff
@@ -130,53 +138,58 @@ public void Remove(string key)
         /// <returns></returns>
         public async Task<bool> TakeLock(string key, Guid lockToken, double milliseconds)
         {
-            var retryCount = 3;
+            var maxRetryAttempt = 3;
             var currentRetry = 0;
             var db = GetDatabase();
 
             while(true)
             {
-                try
-                {        
-                    var transac = db.CreateTransaction();
-                    transac.AddCondition(Condition.KeyExists(key));
+                currentRetry++;
 
-                    var result = await transac.LockTakeAsync(key, lockToken.ToString(), TimeSpan.FromMilliseconds(milliseconds)).ConfigureAwait(false);
-
-                    if (result)
+                try
+                {
+                    var redisLockKey = string.Format("lock_{0}", key);
+                    
+                    var trans = db.CreateTransaction();
+                    trans.AddCondition(Condition.KeyNotExists(redisLockKey));
+                    var acquireLockTask = trans.LockTakeAsync(redisLockKey, lockToken.ToString(), TimeSpan.MaxValue).ConfigureAwait(false);
+                    var setExpiryTimeTask = trans.KeyExpireAsync(redisLockKey, TimeSpan.FromMilliseconds(milliseconds)).ConfigureAwait(false);
```

Comment:
As per conversation. If this line is here only because of a Redis client bug. Please add a code comment referencing the bug, so that we can track and remove the line when the bug is fixed.